docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -116,37 +265,42 @@ Remember, anything can be a sink, even the index of an array access:
 pattern-sinks:
   - patterns:
     - pattern-inside: $ARRAY[$SINK]
-    - pattern: $SINK
+    - focus-metavariable: $SINK
 ```
 
 This way we tell Semgrep that we do not want arrays to be accessed with tainted indexes.
 
-### Sanitized by side-effect
+Minimizing false positives
+--------------------------
 
-Typically a sanitizer will be some function that gets tainted data and returns untainted one. But it does not need to be that way. Sometimes data gets sanitized via side-effect, and taint mode can handle this too, for example:
+Since taint mode is intra-procedural, it does not know what other functions do. By default, Semgrep assumes that taint could always propagate through other functions.
+
+For example, in the code below, `some_safe_function` receives tainted data as input and, to be on the safe side, Semgrep assumes that it will also return tainted data as output. As a result, a finding is produced.
 
 ```javascript
-pattern-sanitizers:
-  - patterns:
-    - pattern-inside: |
-        $JWT.verify($TOKEN, ...)
-        ...
-    - pattern: $TOKEN
+var x = some_safe_function(tainted);
+sink(x);
 ```
 
-This example just annotates as sanitized all the occurrences of a `$TOKEN` that happen after calling `verify` on it.
+In some codebases, this conservative assumption may produce too many false positives. If that is the case, you could first consider enumerating such functions as sanitizers:
 
-You can also use the presence of (for example) an  if conditional as a sanitizer:
-
-```javascript
+```yaml
 pattern-sanitizers:
-  - patterns:
-    - pattern-inside: |
-        if !strings.HasPrefix($PATH, <... $TARGET ...>, ...) {...}
-        ...
-    - pattern: $PATH
+- pattern: some_safe_function(...)
 ```
 
+If this is too cumbersome, then you can easily make Semgrep assume the opposite, that is, to assume that functions are safe by default. This is done by setting `taint_assume_safe_functions: true` under the rule [`options`](/writing-rules/rule-syntax/#options) key. If you use this approach, you instead have to enumerate your taint propagators:
+
+<iframe src="https://semgrep.dev/embed/editor?snippet=gBD0" border="0" frameBorder="0" width="100%" height="432"></iframe>
+
+Similarly, by default, Semgrep assumes that indexing an array with a tainted index (that is, `array[tainted]`) is a tainted expression, even if the array itself is not tainted. You can set `taint_assume_safe_indexes: true` to instead assume that these expressions are safe.
+
+If you want to have very fine-grained control over how taint is propagated, you can set `taint_only_propagate_through_assignments: true`, and then Semgrep will only propagate taint via trivial assignemnts of the form `x = tainted`. You would need to specify taint propagators for anything else.
```

Comment:
What do we mean by "anything else" here? @IagoAbal

Reply:
Any other expression that acts as a taint propagator. For example, `x = taint_propagator(tainted)` or `x = tainted + "foo"` won't propagate any taint from `tainted` to `x` unless you explicitly say so, whereas in the default setting this happens automatically.

Reply:
I will push in a commit (as the diff here is outdated, no suggestion is possible), and substitute the text with the following (just a small change):

```
In this case, you need to specify taint propagators or any other expression that can act as a taint propagator.
```

Reply:
@IagoAbal I will push in a commit (as the diff here is outdated, no suggestion is possible), and substitute the text with the following (just a small change):

```md
If you want to have very fine-grained control over how taint is propagated, set `taint_only_propagate_through_assignments: true`, and then Semgrep only propagates taint through trivial
assignments of the form `x = tainted`. In this case, you need to specify taint propagators or any other expression that can act as a taint propagator. For example, `x = taint_propagator(tainted)` or `x = tainted + "foo"` does not
propagate any taint from `tainted` to `x` unless it is explicitly defined, whereas in this happens automatically in the default taint analysis setting.
```

Reply:
If this is misleading, feel free to rewrite this as you find best!